File: javaapp/src/main/java/com/javapp/cryptography/SecretHandshake/HandshakeCalculator.java
package com.javapp.cryptography.SecretHandshake;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HandshakeCalculator {

    private static final int REVERSE_SIGNAL_BIT_POSITION = 4;

    List<Signal> calculateHandshake(final int number){
        final List<Signal> result = new ArrayList<>();
        for(Signal signal: Signal.values()){
            if( isBitSet(signal.ordinal(), number)){
                result.add(signal);
            }
        }

        if(isBitSet(REVERSE_SIGNAL_BIT_POSITION, number)){
            Collections.reverse(result);
        }

        return result;
    }

    private boolean isBitSet(final int position, int number){
        return ((number >> position) & 1) == 1;
    }
}

File: app/src/main/kotlin/com/kotlinground/puzzles/secrethandshake/HandshakeCalculator.kt
package com.kotlinground.puzzles.secrethandshake

/**
 * @author lusinabrian on 15/01/18.
 */
object HandshakeCalculator {

    private const val REVERSE_POSITION = 4

    fun calculateHandshake(decimalNumber : Int) : List<Signal>{
        val result = Signal.values().filter { isBitSet(it.ordinal, decimalNumber) }

        return if(isBitSet(REVERSE_POSITION, decimalNumber)) result.reversed() else result
    }

    private fun isBitSet(position : Int, decimalNumber: Int) : Boolean {
        return decimalNumber shr position and 1 == 1
    }

}